### ISSUE TITLE
AttributeError: module 'inspect' has no attribute 'getargspec'

### ISSUE THREAD

**Issue by v-soujanya** [python/cpython#100680]
we are working on Azure devops CLI extension, recently we have encountered some issues during build pipeline run. Pipeline run is failing due to the below error:

```pytb
__________ AdminBannerTests.test_admin_banner_addUpdateShowListRemove __________
/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/azure_devtools/scenario_tests/decorators.py:38: in _preparer_wrapper
    trim_kwargs_from_test_function(fn, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fn = <function disable_telemetry.<locals>.wrapper at 0x7fc5d610ede0>
kwargs = {}

    def trim_kwargs_from_test_function(fn, kwargs):
        # the next function is the actual test function. the kwargs need to be trimmed so
        # that parameters which are not required will not be passed to it.
        if not is_preparer_func(fn):
>           args, _, kw, _ = inspect.getargspec(fn)  # pylint: disable=deprecated-method
E           AttributeError: module 'inspect' has no attribute 'getargspec'
```

we have gone through the pyhton 3.11 doc https://docs.python.org/3.11/library/inspect.html#inspect.getargspec and "getargspec" is deprecated. can someone provide the solution on it?

**Comment by sku2000** [comment]
Maybe you can use inspect.getfullargspec(func). There are discussions on this [issues](https://github.com/lmfit/lmfit-py/issues/397)

**Comment by hugovk** [comment]
`inspect.getargspec()` was deprecated in Python 3.0 (released in 2008), raised a `DeprecationWarning` since 3.5 (2015) and removed in 3.11 (2022). The upgrade advice is to use [`inspect.signature()`](https://docs.python.org/3.11/library/inspect.html#inspect.signature) or [`inspect.getfullargspec()`](https://docs.python.org/3.11/library/inspect.html#inspect.getfullargspec) instead.

See:

* https://docs.python.org/3.11/whatsnew/3.11.html#removed
* https://docs.python.org/3.5/library/inspect.html#inspect.getargspec
* https://github.com/python/cpython/issues/89483
* https://github.com/python/cpython/issues/64637

**Comment by v-soujanya** [comment]
@xandrade Thank you for the info! Yes, I checked the discussion, now my concern is how to use this inspect.getfullargsspec(), where to change it. The below is the function it was throwing error.

def disable_telemetry(test_function):
    def wrapper(*args, **kwargs):
        print("Disabling Telemetry.")
        os.environ["AZURE_CORE_COLLECT_TELEMETRY"] = "no"
        test_function(*args, **kwargs)
    return wrapper
Could you please help here, which are the files need to changes to use inspect.getfullargsspec()?

**Comment by hugovk** [comment]
You're calling something in https://github.com/Azure/azure-python-devtools which doesn't support Python 3.11 and is no longer maintained by Microsoft:

> _Disclaimer:_ The package "azure-devtools" is no longer maintened nor used in any Microsoft projects anymore.

https://github.com/Azure/azure-python-devtools

This happened in December 2021: https://github.com/Azure/azure-python-devtools/commit/48e6798686fb06cbc5f788b997c23ae735ba2892

A quick fix is to use Python 3.10 or earlier, a longer term fix is to replace your use of azure-devtools with something that's still supported. (Alternatively, you could fork and fix it, but it may be better to find a replacement.)

Anyway, this issue tracker is for the CPython language itself, so let's close this issue because it's about third-party tools. You may be better off asking on Stack Overflow or Azure forums for help here.

**Comment by polsala** [comment]
There is a workaround monkeypatching inspect to restore `getargspec` globally
```python
    class Inspect:
        @staticmethod
        def monkeypatch():
            import inspect
            print('Applying monkeypatch for inspect')
            if hasattr(inspect, 'getfullargspec') and not hasattr(inspect, 'getargspec'):
                def getargspec(func):
                    from collections import namedtuple
                    inspect_result = inspect.getfullargspec(func)
                    ArgSpec = namedtuple('ArgSpec', 'args varargs keywords defaults')
                    return ArgSpec(
                        inspect_result.args, inspect_result.varargs,
                        inspect_result.varkw, inspect_result.defaults
                    )
                inspect.getargspec = getargspec

Inspect.monkeypatch()
```

![image](https://github.com/python/cpython/assets/15796004/2d79e0c7-7023-431d-adb7-488fce77ae31)
